### DISCUSSION TITLE
DialogService, DynamicDialogConfig

### DISCUSSION THREAD

**Discussion by ibrahimkelly** [primefaces/community#2615] [PrimeNG] [answered]
Hi 👋 everyone 

1. In the documentation, DynamicDialogConfig example is missing, for retrieving data from parent component.

Data is undefined in my child component using the standalone components.

**Comment by qwadrox** [comment]
Hi @ibrahimkelly 

1. You can start by trying to ensure correct type safeness by using the mention DynamicDialogConfig class in the parent component, which is a generic class:

```ts
// interface
export interface IDialogData {
  productName: string;
}

this.ref = this.dialogService.open(ProductListDemo, <
  DynamicDialogConfig<IDialogData>
>{
  data: {
    productName: 'Something',
  },
  header: 'Select a Product',
  width: '50vw',
  contentStyle: { overflow: 'auto' },
  breakpoints: {
    '960px': '75vw',
    '640px': '90vw',
  },
  templates: {
    footer: Footer,
  },
});
```
2. Access the data in the Child Component

```ts
data!: IDialogData;

constructor(
    public config: DynamicDialogConfig<IDialogData>
) {}

ngOnInit() {
    this.data = this.config.data;
}
```
_This approach guarantees that the data passed between components follows to the specified type._

Indeed there was a problem with the provided example, I forked it and included the example above. You can check it for now: https://stackblitz.com/edit/e1msjd?file=src%2Fapp%2Fdemo%2Fproductlistdemo.ts

**Reply by ibrahimkelly** [reply]
> Hi @ibrahimkelly 
> 
> 1. You can start by trying to ensure correct type safeness by using the mention DynamicDialogConfig class in the parent component, which is a generic class:
> 
> ```ts
> // interface
> export interface IDialogData {
>   productName: string;
> }
> 
> this.ref = this.dialogService.open(ProductListDemo, <
>   DynamicDialogConfig<IDialogData>
> >{
>   data: {
>     productName: 'Something',
>   },
>   header: 'Select a Product',
>   width: '50vw',
>   contentStyle: { overflow: 'auto' },
>   breakpoints: {
>     '960px': '75vw',
>     '640px': '90vw',
>   },
>   templates: {
>     footer: Footer,
>   },
> });
> ```
> 2. Access the data in the Child Component
> 
> ```ts
> data!: IDialogData;
> 
> constructor(
>     public config: DynamicDialogConfig<IDialogData>
> ) {}
> 
> ngOnInit() {
>     this.data = this.config.data;
> }
> ```
> _This approach guarantees that the data passed between components follows to the specified type._
> 
> Indeed there was a problem with the provided example, I forked it and included the example above. You can check it for now: https://stackblitz.com/edit/e1msjd?file=src%2Fapp%2Fdemo%2Fproductlistdemo.ts
> 

Gonna try it and will be back soon, thanks in advance.

**Reply by ibrahimkelly** [reply]
> Hi @ibrahimkelly
> 
> 1. You can start by trying to ensure correct type safeness by using the mention DynamicDialogConfig class in the parent component, which is a generic class:
> 
> ```ts
> // interface
> export interface IDialogData {
>   productName: string;
> }
> 
> this.ref = this.dialogService.open(ProductListDemo, <
>   DynamicDialogConfig<IDialogData>
> >{
>   data: {
>     productName: 'Something',
>   },
>   header: 'Select a Product',
>   width: '50vw',
>   contentStyle: { overflow: 'auto' },
>   breakpoints: {
>     '960px': '75vw',
>     '640px': '90vw',
>   },
>   templates: {
>     footer: Footer,
>   },
> });
> ```
> 
> 2. Access the data in the Child Component
> 
> ```ts
> data!: IDialogData;
> 
> constructor(
>     public config: DynamicDialogConfig<IDialogData>
> ) {}
> 
> ngOnInit() {
>     this.data = this.config.data;
> }
> ```
> 
> _This approach guarantees that the data passed between components follows to the specified type._
> 
> Indeed there was a problem with the provided example, I forked it and included the example above. You can check it for now: https://stackblitz.com/edit/e1msjd?file=src%2Fapp%2Fdemo%2Fproductlistdemo.ts


Hi @qwadrox , I saw the yours but there are a bit differents on implementation.
I made a simple example on StackBlitz here is the link below, check it please.
https://stackblitz.com/edit/stackblitz-starters-ncw1gt?file=src%2Fapp%2Fchild%2Fchild.component.html

Thanks 👍🏾

**Reply by qwadrox** [reply, marked as answer]
The data is undefined in the ChildComponent because you're providing DynamicDialogConfig in both the parent (App) and child (ChildComponent) components. To fix:

You don't need to provide DynamicDialogConfig in either component because:

- The parent doesn't use it directly.
- The child receives it automatically from the dialog service. ChildComponent is created dynamically by the dialog service, and there it is being injected using an injector. [const componentRef = createComponent(DynamicDialogComponent, { environmentInjector: this.appRef.injector, elementInjector: new DynamicDialogInjector(this.injector, map) });](https://github.com/primefaces/primeng/blob/1f6c5d95271e5b48aabdcd15bc040a1f3bea5adc/src/app/components/dynamicdialog/dialogservice.ts#L46-L62)

Here is the revised example: https://stackblitz.com/edit/stackblitz-starters-hrec74?file=src%2Fapp%2Fchild%2Fchild.component.ts

**Reply by ibrahimkelly** [reply]
Bro, yours review is working perfectly 👍🏾, but in my code base the error persiste, I applied your changes still remaining.
I can make screen share on Discord if you get time.

Thanks 👍🏾

**Reply by ibrahimkelly** [reply]
> The data is undefined in the ChildComponent because you're providing DynamicDialogConfig in both the parent (App) and child (ChildComponent) components. To fix:
> 
> You don't need to provide DynamicDialogConfig in either component because:
> 
> * The parent doesn't use it directly.
> * The child receives it automatically from the dialog service. ChildComponent is created dynamically by the dialog service, and there it is being injected using an injector. [const componentRef = createComponent(DynamicDialogComponent, { environmentInjector: this.appRef.injector, elementInjector: new DynamicDialogInjector(this.injector, map) });](https://github.com/primefaces/primeng/blob/1f6c5d95271e5b48aabdcd15bc040a1f3bea5adc/src/app/components/dynamicdialog/dialogservice.ts#L46-L62)
> 
> Here is the revised example: https://stackblitz.com/edit/stackblitz-starters-hrec74?file=src%2Fapp%2Fchild%2Fchild.component.ts

Thank you so much for your assistance 🥇 .
